File: app/src/main/java/cn/blinkdagger/androidLab/utils/AppUtil.java
package cn.blinkdagger.androidLab.utils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.text.TextUtils;


/**
 * Created by d on 2016/11/16.
 * APP信息辅助类
 */

public class AppUtil {

    /**
     * 获取包名
     *
     * @param context 上下文
     * @return 本APP的包名
     */
    public static String getPackageName(Context context) {
        return context.getPackageName();
    }

    /**
     * 获取App名称
     *
     * @param context 上下文
     * @return 本App名称
     */
    public static String getAppName(Context context) {
        return getAppName(context, context.getPackageName());
    }

    /**
     * 获取App名称
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App名称
     */
    public static String getAppName(Context context, String packageName) {
        if (TextUtils.isEmpty(packageName)) return null;
        try {
            PackageManager pm = context.getPackageManager();
            PackageInfo pi = pm.getPackageInfo(packageName, 0);
            return pi == null ? null : pi.applicationInfo.loadLabel(pm).toString();
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 获取版本号
     *
     * @param context 上下文
     * @return 本APP的版本号
     */
    public static int getVersionCode(Context context) {
        return getVersionCode(context, context.getPackageName());
    }

    /**
     * 获取App版本号
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App版本号
     */
    public static int getVersionCode(Context context, String packageName) {
        if (TextUtils.isEmpty(packageName)) return -1;
        try {
            PackageManager pm = context.getPackageManager();
            PackageInfo pi = pm.getPackageInfo(packageName, 0);
            return pi == null ? -1 : pi.versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return -1;
        }
    }

    /**
     * 获取版本名称
     *
     * @param context 上下文
     * @return 本APP的版本名称
     */
    public static String getVersionName(Context context) {
        return getVersionName(context, context.getPackageName());
    }

    /**
     * 获取App版本名称
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App版本名称
     */
    public static String getVersionName(Context context, String packageName) {
        if (TextUtils.isEmpty(packageName)) return null;
        try {
            PackageManager pm = context.getPackageManager();
            PackageInfo pi = pm.getPackageInfo(packageName, 0);
            return pi == null ? null : pi.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return null;
        }
    }


    /**
     * 判断App是否是Debug版本
     * （注意混淆的时候排除相应的BuildConfig类 ）
     *
     * @param packageName 包名
     * @return boolean
     */
    public static boolean isAppDebug(String packageName) {
        if (TextUtils.isEmpty(packageName)) return false;
        try {
            Class clazz = Class.forName(packageName + ".BuildConfig");
            Boolean isDebug = clazz.getField("DEBUG").getBoolean(clazz);
            return isDebug == null ? false : isDebug;
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            return false;
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
            return false;
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return false;
        }
    }


    /**
     * 进入App具体设置
     *
     * @param context 上下文
     */
    public static void getAppDetailsSetting(Context context) {
        getAppDetailsSetting(context, context.getPackageName());
    }

    /**
     * 进入App具体设置
     *
     * @param context     上下文
     * @param packageName 包名
     */
    public static void getAppDetailsSetting(Context context, String packageName) {
        if (TextUtils.isEmpty(packageName)) return;
        context.startActivity(IntentUtil.getAppDetailsSettingIntent(packageName));
    }

    /**
     * 进入系统设置
     *
     * @param context     上下文
     * @param packageName 包名
     */
    public static void getSystemSetting(Context context, String packageName) {
        if (TextUtils.isEmpty(packageName)) return;
        context.startActivity(IntentUtil.getAppDetailsSettingIntent(packageName));
    }
}

File: app/src/main/java/cn/blinkdagger/androidLab/utils/AppUtil.kt
package cn.blinkdagger.androidLab.utils

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.text.TextUtils

/**
 * Created by d on 2016/11/16.
 * APP信息辅助类
 */
object AppUtil {
    /**
     * 获取包名
     *
     * @param context 上下文
     * @return 本APP的包名
     */
    fun getPackageName(context: Context): String {
        return context.packageName
    }

    /**
     * 获取App名称
     *
     * @param context 上下文
     * @return 本App名称
     */
    fun getAppName(context: Context): String? {
        return getAppName(context, context.packageName)
    }

    /**
     * 获取App名称
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App名称
     */
    fun getAppName(context: Context, packageName: String?): String? {
        return if (TextUtils.isEmpty(packageName)) null else try {
            val pm = context.packageManager
            val pi = pm.getPackageInfo(packageName, 0)
            pi?.applicationInfo?.loadLabel(pm)?.toString()
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 获取版本号
     *
     * @param context 上下文
     * @return 本APP的版本号
     */
    fun getVersionCode(context: Context): Int {
        return getVersionCode(context, context.packageName)
    }

    /**
     * 获取App版本号
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App版本号
     */
    fun getVersionCode(context: Context, packageName: String?): Int {
        return if (TextUtils.isEmpty(packageName)) -1 else try {
            val pm = context.packageManager
            val pi = pm.getPackageInfo(packageName, 0)
            pi?.versionCode ?: -1
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            -1
        }
    }

    /**
     * 获取版本名称
     *
     * @param context 上下文
     * @return 本APP的版本名称
     */
    fun getVersionName(context: Context): String? {
        return getVersionName(context, context.packageName)
    }

    /**
     * 获取App版本名称
     *
     * @param context     上下文
     * @param packageName 包名
     * @return App版本名称
     */
    fun getVersionName(context: Context, packageName: String?): String? {
        return if (TextUtils.isEmpty(packageName)) null else try {
            val pm = context.packageManager
            val pi = pm.getPackageInfo(packageName, 0)
            pi?.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 判断App是否是Debug版本
     * （注意混淆的时候排除相应的BuildConfig类 ）
     *
     * @param packageName 包名
     * @return boolean
     */
    fun isAppDebug(packageName: String): Boolean {
        return if (TextUtils.isEmpty(packageName)) false else try {
            val clazz = Class.forName("$packageName.BuildConfig")
            val isDebug = clazz.getField("DEBUG").getBoolean(clazz)
            isDebug ?: false
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
            false
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
            false
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
            false
        }
    }

    /**
     * 进入App具体设置
     *
     * @param context 上下文
     */
    fun getAppDetailsSetting(context: Context) {
        getAppDetailsSetting(context, context.packageName)
    }

    /**
     * 进入App具体设置
     *
     * @param context     上下文
     * @param packageName 包名
     */
    fun getAppDetailsSetting(context: Context, packageName: String) {
        if (TextUtils.isEmpty(packageName)) return
        context.startActivity(IntentUtil.getAppDetailsSettingIntent(packageName))
    }

    /**
     * 进入系统设置
     *
     * @param context     上下文
     * @param packageName 包名
     */
    fun getSystemSetting(context: Context, packageName: String) {
        if (TextUtils.isEmpty(packageName)) return
        context.startActivity(IntentUtil.getAppDetailsSettingIntent(packageName))
    }

    /**
     * 　获取渠道号
     *
     * @return
     */
    fun getChannelName(context: Context, channelMetaNameKey: String?): String? {
        var channelName: String? = null
        val appInfo: ApplicationInfo
        try {
            appInfo = context.packageManager.getApplicationInfo(getPackageName(context), PackageManager.GET_META_DATA)
            channelName = appInfo.metaData.getString(channelMetaNameKey)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        return channelName
    }
}